File: Lab3/src/KI302/Melnyk/Lab3/MobilePhone.java
package KI302.Melnyk.Lab3;

import java.io.FileNotFoundException;

/**
 * Class MobilePhone represents a specific type of telephone that extends the abstract Telephone class
 * and implements the Camera interface.
 *
 * @author Vladyslav Melnyk
 * @version 1.0
 * @since version 2.0
 */
public class MobilePhone extends Telephone implements Camera {
    
    private boolean hasTouchScreen;

    public MobilePhone() throws FileNotFoundException {
        super();
        this.hasTouchScreen = true;
        log("MobilePhone default constructor called.");
    }

    public MobilePhone(String modelName, Screen screen, Battery battery, SimCard simCard, boolean hasTouchScreen) throws FileNotFoundException {
        super(modelName, screen, battery, simCard);
        this.hasTouchScreen = hasTouchScreen;
        log("MobilePhone parameterized constructor called.");
    }

    @Override
    public String getConnectionType() {
        return "Mobile Network (4G/5G)";
    }
    
    /**
     * Implementation of the takePhoto method from the Camera interface.
     * Simulates taking a photo.
     */
    @Override
    public void takePhoto() {
        log("Taking a photo with the mobile phone's camera.");
        System.out.println("Click! Photo taken.");
    }

    public void connectToWifi(String networkName) {
        log("Connecting to Wi-Fi network: " + networkName);
        System.out.println("Successfully connected to " + networkName);
    }

    @Override
    public void getFullPhoneInfo() {
        super.getFullPhoneInfo();
        System.out.println("Touch Screen: " + (hasTouchScreen ? "Yes" : "No"));
        log("Displayed full info for MobilePhone.");
    }
}

File: Lab9/phone_package/mobile_phone.py
# mobile_phone.py
from .telephone import Telephone

class MobilePhone(Telephone):
    """
    Клас-нащадок, що представляє мобільний телефон.
    Успадковує клас Telephone і додає нові можливості.
    """
    
    # Конструктор класу MobilePhone
    def __init__(self, model, has_touchscreen):
        # Викликаємо конструктор батьківського класу Telephone
        super().__init__(model)
        self._has_touchscreen = has_touchscreen

    def connect_to_wifi(self, network_name):
        """Новий метод для підключення до Wi-Fi."""
        if self._is_on:
            print(f"{self._model}: Підключено до Wi-Fi '{network_name}'.")
        else:
            print(f"{self._model}: Неможливо підключитись до Wi-Fi, телефон вимкнено.")

    # Перевизначаємо метод батьківського класу, щоб додати нову інформацію
    def get_info(self):
        """Повертає розширену інформацію про мобільний телефон."""
        # Викликаємо оригінальний метод get_info() з класу Telephone
        base_info = super().get_info()
        touch_info = "Так" if self._has_touchscreen else "Ні"
        return f"{base_info}, Сенсорний екран: {touch_info}"